### DISCUSSION TITLE
How to do graphql queries over local files?

### DISCUSSION THREAD

**Discussion by bjdmeest** [comunica/comunica#631] [Q&A] [answered]
#### Issue type:

- :question: question

____
#### Question:

There's  a sparql-init for remote files for both sparql and graphql, and a sparql-file for local files for sparql, is it possible to expand on the latter one to also allow graphql?

**Comment by rubensworks** [comment]
It should theoretically be possible to follow [the GraphQL-LD example](https://github.com/comunica/comunica/tree/master/packages/actor-init-sparql#optional-graphql-ld), and swap `@comunica/actor-init-sparql` with `@comunica/actor-init-sparql-file`.

Let me know if that doesn't work, and we'll make this a feature request.

In any case, I see @pheyvaer made this, so that should solve your problem in any case: https://www.npmjs.com/package/graphqlld-on-file

**Comment by pheyvaer** [comment]
Note: my code should be updated to the latest version of GraphQL-LD though.

**Comment by bjdmeest** [comment]
If I try the following, I get the error "This actor can only parse SPARQL queries"

```js
const newEngine = require('@comunica/actor-init-sparql-file').newEngine;
const bindingsStreamToGraphQl = require('@comunica/actor-sparql-serialize-tree').bindingsStreamToGraphQl;
const fs = require('fs');

console.log('starting');
const myEngine = newEngine();
const config = {
  sources: ['http://knows.proxy.ilabt.imec.be/person/adimou/#me','https://ruben.verborgh.org/profile/#me', 'http://knows.proxy.ilabt.imec.be/person/office/master-theses.json'],
  queryFormat: 'graphql',
  "@context": JSON.parse(fs.readFileSync('./config-with-context.json', 'utf8'))['@context']
};
console.log('this config ' + JSON.stringify(config));
myEngine.query(fs.readFileSync('./query.graphql', 'utf8'), config)
  .then(function (result) { return bindingsStreamToGraphQl(result.bindingsStream, result.context); })
  .then(console.log);
```

Also, probably a next question of me will be: how to query both local _and_ online sources? :)

**Comment by rubensworks** [comment]
> If I try the following, I get the error "This actor can only parse SPARQL queries"

Hmm.
This means that the GraphQL actor(s) are not configured for @comunica/actor-init-sparql-file.

It should be easy to add this though.
If you're up for it, you can submit a PR where [this line](https://github.com/comunica/comunica/blob/master/packages/actor-init-sparql/config/config-default.json#L20) is added to the [@comunica/actor-init-sparql-file config](https://github.com/comunica/comunica/blob/master/packages/actor-init-sparql-file/config/config-default.json).

> Also, probably a next question of me will be: how to query both local and online sources? :)

This would require creating your own Comunica configuration,
where you basically create something like @comunica/actor-init-sparql-somethingcustom.

The easiest way to get started with this, is by copying [the actor-init-sparql-file package](https://github.com/comunica/comunica/tree/master/packages/actor-init-sparql-file), and adding all entries you want to its config.

**Comment by rubensworks** [comment, marked as answer]
I just created a dedicated [Comunica examples repo](https://github.com/comunica/examples), which currently contains one example of creating a custom Comunica engine with a custom actor.

Feel free to open an example request issue there if things should not be sufficiently clear with that one example.
